Reworking of userspace tools (rapiddisk and rapiddiskd)

matteotenca opened the pull request:
Goals achieved:

1. Fix memory leaks
2. Implement a better fork procedure
3. Rewrite the `switch` logic in `main.c`
4. Rewrite the `if/then/else` logic in `net.c`
5. Rewrite the daemon main loop and implement a clean exit logic based on signals
6. Sharing of JSON output functions between command and daemon
7. Get rid of calls to external executables: the daemon does not invoke `rapiddisk` any more and both use `libdevmapper` directly, without invoking `dmsetup`
8. All functions which outputs a message in `rapiddisk` now can print it in JSON too
9. New `Makefile`s with some new targets
10. Added the `-d` switch to the daemon to avoid forking and forcing it to print messages to STDERR too
11. Added documentation in `doxygen` format

This PR includes all the commits in PR #132, which can be closed.

Regards

pkoutoupis commented on src/common.h:
```diff
@@ -50,10 +56,11 @@
-#define WRITETHROUGH		0
-#define WRITEAROUND		1
-#define WRITEBACK		2
+//#define WRITETHROUGH		0
```
We can remove these commented out lines.

matteotenca replied:
> We can remove these commented out lines.

Fixed in forthcoming commit, FYI these values are currently set at lines:

https://github.com/pkoutoupis/rapiddisk/blob/db51403ad0934462e8cd710da12bd56879808596/src/common.h#L176-L180

pkoutoupis commented on src/json.c:
```diff
@@ -28,30 +28,68 @@
  ********************************************************************************/
 
 #include "common.h"
+//#ifdef SERVER
```
We can remove these commented lines.

matteotenca replied:
Fixed.

pkoutoupis commented on src/json.c:
```diff
@@ -28,30 +28,68 @@
+ * This function is intended to provide a simple way to create a JSON string which contains
+ * a status ("Success" or "Failed") derived from an integer, and an optional message.
+ *
+ * \param return_value if @p return_value is != @p 'SUCCESS', the JSON result will
```
Is this comment supposed to be @param and not \param?

matteotenca replied:
Both syntax are ok, but I changed the `\` prefix to the `@` prefix because I (hope to have) always used `@` elsewhere.

Another use of `\` is at line:

https://github.com/pkoutoupis/rapiddisk/blob/db51403ad0934462e8cd710da12bd56879808596/src/json.c#L70

Both fixed in forthcoming commit.

pkoutoupis commented on src/json.c:
```diff
@@ -395,15 +529,26 @@ int json_nvmet_view_exports(struct NVMET_PROFILE *nvmet, struct NVMET_PORTS *por
-	json_decref(ports_array);
-	json_decref(array);
+	char *jdumped = json_dumps(root, 0);
+//	if (jdumped != NULL) {
```
Can we remove this commented code? It looks as if it was only used for troubleshooting.

matteotenca replied:
Yes, fixed!

pkoutoupis commented on src/json.c:
```diff
@@ -459,38 +621,35 @@ int json_nvmet_view_ports(struct NVMET_PORTS *ports)
+			*json_str = jdumped;
+			rc = SUCCESS;
+		}
+//		sprintf(message, "%s\n", jdumped);
```
Again,  it looks aa if this was used for troubleshooting and if we do not need it, we should remove the commented code.

matteotenca replied:
Removed, fixed.

pkoutoupis commented on src/json.h:
```diff
@@ -0,0 +1,24 @@
+/** @file
```
Can we follow the style of heading for new files as seen in others? Feel free to keep your name and date (maybe add an e-mail if you feel comfortable enough). I just ask that we make all file headings look consistent.

matteotenca replied:
Fixed!

pkoutoupis commented on src/main.h:
```diff
@@ -0,0 +1,60 @@
+#define ERR_NO_DEVICES		"Unable to locate any RapidDisk devices."
+#define ERR_NO_MEMUSAGE		"Error. Unable to retrieve memory usage data."
+#define ERR_INVALID_PORT	"Error. Invalid port number."
+#define ERR_NO_RDEVICES		"No RapidDisk devices exist on the system."
```
It seems like ERR_NO_RDEVICES serves the same purpose as ERR_NO_DEVICES. I recommend that we use one and stick with it and replace all instance of the removed with the one we decided to keep.

matteotenca replied:
Fixed, I removed `ERR_NO_RDEVICES`.

pkoutoupis commented on src/net.c:
```diff
@@ -1,3 +1,8 @@
+/** @file net.c
```
See my note above on file headings. We should keep things consistent. Is this format for doxygen?

matteotenca replied:
Fixed.

pkoutoupis commented on src/net.c:
```diff
@@ -45,278 +54,574 @@ static enum MHD_Result answer_to_connection(void *cls, struct MHD_Connection *co
+	}
+	while (!server_stop_requested) {
+		/* On any signal, sleep is always interrupted and this allows to the signal handler function to be invoked */
+		sleep(120 * 4);
```
Just a question: is there any significance to this value? 120 * 4

pkoutoupis commented on src/nvmet.c:
```diff
@@ -595,25 +951,36 @@ int nvmet_unexport_volume(unsigned char *device, unsigned char *host, int port)
+	 * because prevented the complete removal of the mapping
+	 */
+
+//	if ((strlen(host) != 0) && (port != INVALID_VALUE)) {
```
I need to revisit this and figure out what I did what I did here and the problem I was intending to solve.

matteotenca replied:
Ok!

pkoutoupis commented on src/nvmet.c:
```diff
@@ -739,59 +1055,94 @@ int nvmet_view_ports(bool json_flag)
+
+	ports = nvmet_scan_ports(error_message);
+	if ((ports == NULL) && (strlen(error_message) != 0)) {
+//		print_message(INVALID_VALUE, error_message, json_flag);
```
If we do not need this commented code here, please remove.

pkoutoupis commented on src/nvmet.c:
```diff
@@ -739,59 +1055,94 @@ int nvmet_view_ports(bool json_flag)
-	sprintf(ip, "%s", nvmet_interface_ip_get(interface));
+	char *ipaddr = nvmet_interface_ip_get(interface, return_message);
+	if (ipaddr == NULL) {
+//		msg = "Error. IP address %s is invalid.";
```
Why do we not want to return a string message here?

matteotenca replied:
Fixed!

pkoutoupis commented on src/nvmet.c:
```diff
@@ -802,58 +1153,133 @@ int nvmet_enable_port(unsigned char *interface, int port, int protocol)
+
+	nvmet = nvmet_scan_subsystem(error_message);
+	if ((nvmet == NULL) && (strlen(error_message) != 0)) {
+//		print_message(INVALID_VALUE, error_message, json_flag);
```
Is this something we should enable for enabled verbosity or should we remove commented code?

matteotenca replied:
Removed commented code.

pkoutoupis commented on src/nvmet.c:
```diff
@@ -802,58 +1153,133 @@ int nvmet_enable_port(unsigned char *interface, int port, int protocol)
+
+	ports = nvmet_scan_ports(error_message);
+	if ((ports == NULL) && (strlen(error_message) != 0)) {
+//		print_message(INVALID_VALUE, error_message, json_flag);
```
Same question here as above.

matteotenca replied:
Removed commented code.

pkoutoupis commented on src/nvmet.c:
```diff
@@ -802,58 +1153,133 @@ int nvmet_enable_port(unsigned char *interface, int port, int protocol)
+
+	ports = nvmet_scan_all_ports(error_message);
+	if ((ports == NULL) && (strlen(error_message) != 0)) {
+//		print_message(INVALID_VALUE, error_message, json_flag);
```
Same here.

matteotenca replied:
Removed commented code.

pkoutoupis commented on src/nvmet.h:
```diff
@@ -0,0 +1,31 @@
+/** @file
```
Please see my comments above on file headings.

matteotenca replied:
Fixed!

pkoutoupis commented on src/rapiddiskd.c:
```diff
@@ -0,0 +1,377 @@
+ **
+ ** @project: rapiddisk
+ **
+ ** @filename: daemon.c
```
This file name field should be changed to reflect the actual change.

matteotenca replied:
Fixed.

pkoutoupis commented on src/rapiddiskd.c:
remove commented code.

matteotenca replied:
Done.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -27,473 +27,800 @@
+ * error message will be printed to stdout.
+ * @param ... Values to be inserted.
+ */
+void print_error(char *format_string, char *return_message, ...) {
```
We can leave it as print_error but seeing how it doesn't just print errors, would it make more sense to rename? If not, it is okay, we can keep it as print_error.

matteotenca replied:
You're right, it's not a good name at all... any suggestion?

pkoutoupis commented on src/rdsk.c:
```diff
@@ -27,473 +27,800 @@
+	char *command = "flush";
+
+	if (!(dmt = dm_task_create(cmdno))) {
+//		printf("%s\n", "dm_task_create error");
```
We should clean up and/or remove the commented code here and below and in the other functions of this file.

matteotenca replied:
Done.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -27,473 +27,800 @@
+	memset(table, 0x0, BUFSZ);
+	if (cache_mode == WRITEBACK)    /* very dangerous mode */
+//		sprintf(string, "echo 0 %llu writecache s %s /dev/%s 4096 0|dmsetup create %s\n",
+//				block_dev_sz, block_dev, ramdisk, name);
```
please remove commented code.

matteotenca replied:
Done.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
-		if ((rc = system(cmd)) != SUCCESS)
-			printf("Unable to flush dirty cache data to %s\n", string);
+		if ((rc = dm_flush_device(string) != SUCCESS)) {
+//		sprintf(cmd, "dmsetup message %s 0 flush\n", string);
```
Please remove commented code.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
-	else {
-		printf("Error. Unable to unmap %s. Please check to make sure "
-		       "nothing is wrong.n\n", string);
+//	sprintf(cmd, "dmsetup remove %s\n", string);
```
Please remove commented code.

matteotenca replied:
Done.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
+	if (rc_prof == NULL) {
+		msg = "No RapidDisk-Cache Mappings exist.";
+		print_error("%s", NULL, msg);
+//		printf("  No RapidDisk-Cache Mappings exist.\n\n");
```
Please remove commented code.

matteotenca replied:
Done.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
+	if (rc_prof == NULL) {
+		msg = "No RapidDisk-Cache Mappings exist.";
+		print_message(rc, msg, TRUE);
+//		printf("  No RapidDisk-Cache Mappings exist.\n\n");
```
Please remove commented code.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
+	if (rc_prof == NULL) {
+		msg = "No RapidDisk-Cache Mappings exist.";
+		print_message(rc, msg, TRUE);
+//		printf("  No RapidDisk-Cache Mappings exist.\n\n");
```
Please remove commented code.

matteotenca replied:
Done.

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
-	return (rc * PAGE_SIZE);
+	}
+	*wc_stats = dm_get_status(cache, WRITEBACK);
+//	free(rc_stats);
```
Please remove commented code.

matteotenca replied:
Fixed!

pkoutoupis commented on src/rdsk.c:
```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
-	if ((fd = open(file, O_WRONLY)) < SUCCESS) {
-		printf("%s: open: %s\n", __func__, strerror(errno));
+	if (rc_prof == NULL) {
+		msg = "No RapidDisk-Cache Mappings exist.";
```
I wonder if we should just predefine these strings in main.h instead of constantly setting msg. It would also reduce extra lines of code in this file.

matteotenca replied:
I did it in a new forthcoming commit, not pushed since not ready. I created `#define`s for many if not all of the error messages specified as static strings, but the work is not finished yet!

pkoutoupis commented on src/rdsk.h:
```diff
@@ -0,0 +1,45 @@
+/** @file
```
Please refer to my notes above on headings.

matteotenca replied:
Fixed.

pkoutoupis commented on src/utils.c:
```diff
@@ -0,0 +1,322 @@
+//
```
See above notes on file headings

matteotenca replied:
Fixed.

pkoutoupis commented on src/utils.h:
```diff
@@ -0,0 +1,20 @@
+/** @file
+ * @brief Utility functions declarations
+ */
```
See above notes on file headings.

matteotenca replied:
Fixed.

pkoutoupis reviewed (changes requested):
So, this is wonderful work. I know that I added a lot of comments but a lot of them are repeats. I still need to checkout and build to run through it in my local set up. The comments today are mostly cosmetic. There may be more coming after I test locally. AGAIN, thank you! I hope to wrap this up very soon to merge into master.

pkoutoupis commented:
OH! I forgot. The RPM spec files need to be updated for the new depends.

matteotenca commented:
> So, this is wonderful work. I know that I added a lot of comments but a lot of them are repeats. I still need to checkout and build to run through it in my local set up. The comments today are mostly cosmetic. There may be more coming after I test locally. AGAIN, thank you! I hope to wrap this up very soon to merge into master.

You're welcome!!

About headers, I rewrote them in a `doxygen` compliant way... maybe they still need some adjustments.

pkoutoupis commented:
Ignore: These are my personal notes...
Update requirements to include libpcre2-dev and libdevmapper-dev.

pkoutoupis commented:
@matteotenca Functionally, this looks great and I cannot seem to find any new memory leaks when stressing the daemon. You did a great job here. Let me do another pass through the new commits and see what is left.

pkoutoupis commented on src/main.h:
```diff
@@ -0,0 +1,62 @@
+#ifndef MAIN_H
+#define MAIN_H
+
+#define ACTION_NONE					0x0
```
Please align all of these values into a nice column.

pkoutoupis commented:
@matteotenca Can you please push your latest commit of changes? Also, does it include the updates to requirements in the spec files?

matteotenca commented:
> @matteotenca Can you please push your latest commit of changes? Also, does it include the updates to requirements in the spec files?

Ok, please give me some days more and I will commit the latest changes!

matteotenca commented:
> @matteotenca Can you please push your latest commit of changes? Also, does it include the updates to requirements in the spec files?

Do you mean in `pkg/rpm`? I give them a try but I encountered some problems creating the rpm with the 'spec.rhel' file under Centos:

```bash
shub@localhost:/tmp/rapiddisk-8.2.0$ rpmlint ~/rpmbuild/SPECS/rapiddisk.spec
/home/shub/rpmbuild/SPECS/rapiddisk.spec: E: specfile-error warning: bogus date in %changelog: Thu May 26 2020 Petros Koutoupis <petros@petroskoutoupis.com>
0 packages and 1 specfiles checked; 1 errors, 0 warnings.
shub@localhost:~/rpmbuild$ rpmbuild  -bb ~/rpmbuild/SPECS/rapiddisk.spec
warning: bogus date in %changelog: Thu May 26 2020 Petros Koutoupis <petros@petroskoutoupis.com>
setting SOURCE_DATE_EPOCH=1654905600
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.WcbelH
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ cd /home/shub/rpmbuild/BUILD
+ cd rapiddisk
/var/tmp/rpm-tmp.WcbelH: line 32: cd: rapiddisk: No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.WcbelH (%prep)


RPM build errors:
    bogus date in %changelog: Thu May 26 2020 Petros Koutoupis <petros@petroskoutoupis.com>
    Bad exit status from /var/tmp/rpm-tmp.WcbelH (%prep)
```

I found out that `May 26` is a Tuesday and not a Thurdsday, I corrected the file but the `%prep` section seems to fail, so I changed it to a simple `%setup -q` but again:

```bash
setting SOURCE_DATE_EPOCH=1654905600
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.9mVhsF
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ cd /home/shub/rpmbuild/BUILD
+ rm -rf rapiddisk-8.2.0
+ /usr/bin/gzip -dc /home/shub/rpmbuild/SOURCES/rapiddisk-8.2.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd rapiddisk-8.2.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.hn4O3z
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ cd rapiddisk-8.2.0
+ make
make -C module
make[1]: Entering directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module'
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
make -C /lib/modules/6.0.7-1.el9.elrepo.x86_64/build M=/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module
make[2]: Entering directory '/usr/src/kernels/6.0.7-1.el9.elrepo.x86_64'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (GCC) 11.2.1 20220127 (Red Hat 11.2.1-9)
  You are using:           gcc (GCC) 11.3.1 20220421 (Red Hat 11.3.1-2)
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.o
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.o
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
  MODPOST /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/Module.symvers
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.mod.o
  LD [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.ko
  BTF [M] /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.ko
Skipping BTF generation for /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.ko due to unavailability of vmlinux
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.mod.o
  LD [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.ko
  BTF [M] /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.ko
Skipping BTF generation for /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.ko due to unavailability of vmlinux
make[2]: Leaving directory '/usr/src/kernels/6.0.7-1.el9.elrepo.x86_64'
make[1]: Leaving directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module'
make -C src
make[1]: Entering directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/src'
Creating Makefile utils.d
Creating Makefile sys.d
Creating Makefile rdsk.d
Creating Makefile rapiddiskd.d
Creating Makefile nvmet.d
Creating Makefile net.d
Creating Makefile main.d
Creating Makefile json.d
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c main.c -o main_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c utils.c -o utils_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c json.c -o json_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c rdsk.c -o rdsk_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c nvmet.c -o nvmet_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c sys.c -o sys_ndebug.o
cc -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o rapiddisk main_ndebug.o utils_ndebug.o json_ndebug.o rdsk_ndebug.o nvmet_ndebug.o sys_ndebug.o  -ljansson -ldevmapper -lpcre2-8
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c rapiddiskd.c -o rapiddiskd_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c json.c -o json-server_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c nvmet.c -o nvmet-server_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c net.c -o net_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c rdsk.c -o rdsk-server_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c sys.c -o sys-server_ndebug.o
cc -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o rapiddiskd rapiddiskd_ndebug.o utils_ndebug.o json-server_ndebug.o nvmet-server_ndebug.o net_ndebug.o rdsk-server_ndebug.o sys-server_ndebug.o  -ljansson -ldevmapper -lpcre2-8 -lmicrohttpd
make[1]: Leaving directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/src'
make -C test
make[1]: Entering directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/test'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/test'
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%install): /bin/sh -e /var/tmp/rpm-tmp.GhIxKX
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ '[' /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64 '!=' / ']'
+ rm -rf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
++ dirname /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ mkdir -p /home/shub/rpmbuild/BUILDROOT
+ mkdir /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ cd rapiddisk-8.2.0
+ rm -rf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ cd src
+ make DESTDIR=/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/ install
Installing all rapiddisk rapiddiskd binary files.
install -D -m 755 -t /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//sbin rapiddisk rapiddiskd
+ cd ../doc
+ make DESTDIR=/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/ install
Nothing to do.
Installing rapiddisk man pages.
mkdir -pv -m 755 /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man/man1
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man/man1'
cp rapiddisk*.1 /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man/man1/
+ cd ../conf
+ make DESTDIR=/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/ install
Nothing to do.
Installing rapiddisk high availability and rapiddiskd daemon files.
install -D -m 755 -t /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//etc/rapiddisk/ rapiddisk.*
install -D -m 755 -t /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//etc/systemd/system rapiddiskd.*
+ cd ../module
+ mkdir -pv /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/'
+ cp -v dkms.conf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
'dkms.conf' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/dkms.conf'
+ cp -v rapiddisk.c rapiddisk-cache.c /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
'rapiddisk.c' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/rapiddisk.c'
'rapiddisk-cache.c' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/rapiddisk-cache.c'
+ cp -v Makefile /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
'Makefile' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/Makefile'
+ /usr/lib/rpm/find-debuginfo.sh -j3 --strict-build-id -m -i --build-id-seed 8.2.0-1 --unique-debug-suffix -8.2.0-1.x86_64 --unique-debug-src-base rapiddisk-8.2.0-1.x86_64 --run-dwz --dwz-low-mem-die-limit 10000000 --dwz-max-die-limit 110000000 --remove-section .gnu.build.attributes -S debugsourcefiles.list /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0
explicitly decompress any DWARF compressed ELF sections in /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/sbin/rapiddisk
explicitly decompress any DWARF compressed ELF sections in /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/sbin/rapiddiskd
extracting debug info from /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/sbin/rapiddisk
extracting debug info from /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/sbin/rapiddiskd
gdb-add-index: No index was created for /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/sbin/rapiddiskd
gdb-add-index: [Was there no debuginfo? Was there already an index?]
gdb-add-index: No index was created for /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/sbin/rapiddisk
gdb-add-index: [Was there no debuginfo? Was there already an index?]
dwz: Too few files for multifile optimization
original debug info size: 40kB, size after compression: 44kB
/usr/lib/rpm/sepdebugcrcfix: Updated 0 CRC32s, 2 CRC32s did match.
+ '[' '%{buildarch}' = noarch ']'
+ QA_CHECK_RPATHS=1
+ case "${QA_CHECK_RPATHS:-}" in
+ /usr/lib/rpm/check-rpaths
+ /usr/lib/rpm/check-buildroot
+ /usr/lib/rpm/redhat/brp-ldconfig
+ /usr/lib/rpm/brp-compress
+ /usr/lib/rpm/redhat/brp-strip-lto /usr/bin/strip
+ /usr/lib/rpm/brp-strip-static-archive /usr/bin/strip
+ /usr/lib/rpm/redhat/brp-python-bytecompile '' 1 0
+ /usr/lib/rpm/brp-python-hardlink
+ /usr/lib/rpm/redhat/brp-mangle-shebangs
*** WARNING: ./etc/systemd/system/rapiddiskd.service is executable but has no shebang, removing executable bit
mangling shebang in /etc/rapiddisk/rapiddisk.sh.rgmanager from /bin/bash to #!/usr/bin/bash
mangling shebang in /etc/rapiddisk/rapiddisk.sh.pacemaker from /bin/bash to #!/usr/bin/bash
Processing files: rapiddisk-8.2.0-1.x86_64
Provides: rapiddisk = 8.2.0-1 rapiddisk(x86-64) = 8.2.0-1
Requires(interp): /bin/sh /bin/sh /bin/sh
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires(post): /bin/sh
Requires(preun): /bin/sh
Requires(postun): /bin/sh
Requires: /usr/bin/bash libc.so.6()(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3)(64bit) libc.so.6(GLIBC_2.34)(64bit) libc.so.6(GLIBC_2.4)(64bit) libc.so.6(GLIBC_2.7)(64bit) libdevmapper.so.1.02()(64bit) libdevmapper.so.1.02(Base)(64bit) libdevmapper.so.1.02(DM_1_02_97)(64bit) libjansson.so.4()(64bit) libjansson.so.4(libjansson.so.4)(64bit) libmicrohttpd.so.12()(64bit) libpcre2-8.so.0()(64bit) rtld(GNU_HASH)
Processing files: rapiddisk-utils-8.2.0-1.x86_64
Provides: rapiddisk-utils = 8.2.0-1 rapiddisk-utils(x86-64) = 8.2.0-1
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires: /usr/bin/bash libc.so.6()(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3)(64bit) libc.so.6(GLIBC_2.34)(64bit) libc.so.6(GLIBC_2.4)(64bit) libc.so.6(GLIBC_2.7)(64bit) libdevmapper.so.1.02()(64bit) libdevmapper.so.1.02(Base)(64bit) libdevmapper.so.1.02(DM_1_02_97)(64bit) libjansson.so.4()(64bit) libjansson.so.4(libjansson.so.4)(64bit) libmicrohttpd.so.12()(64bit) libpcre2-8.so.0()(64bit) rtld(GNU_HASH)
Processing files: rapiddisk-debugsource-8.2.0-1.x86_64
error: Empty %files file /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/debugsourcefiles.list


RPM build errors:
    Empty %files file /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/debugsourcefiles.list
shub@localhost:~/rpmbuild$

```

At the end, I ran:

```bash
shub@localhost:~/rpmbuild$ rpmbuild  --nodebuginfo -bb ~/rpmbuild/SPECS/rapiddisk.spec
setting SOURCE_DATE_EPOCH=1654905600
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.wRoOYu
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ cd /home/shub/rpmbuild/BUILD
+ rm -rf rapiddisk-8.2.0
+ /usr/bin/gzip -dc /home/shub/rpmbuild/SOURCES/rapiddisk-8.2.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd rapiddisk-8.2.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.nbVkyB
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ cd rapiddisk-8.2.0
+ make
make -C module
make[1]: Entering directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module'
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
make -C /lib/modules/6.0.7-1.el9.elrepo.x86_64/build M=/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module
make[2]: Entering directory '/usr/src/kernels/6.0.7-1.el9.elrepo.x86_64'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (GCC) 11.2.1 20220127 (Red Hat 11.2.1-9)
  You are using:           gcc (GCC) 11.3.1 20220421 (Red Hat 11.3.1-2)
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.o
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.o
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD
  MODPOST /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/Module.symvers
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.mod.o
  LD [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.ko
  BTF [M] /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.ko
Skipping BTF generation for /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk-cache.ko due to unavailability of vmlinux
  CC [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.mod.o
  LD [M]  /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.ko
  BTF [M] /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.ko
Skipping BTF generation for /home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module/rapiddisk.ko due to unavailability of vmlinux
make[2]: Leaving directory '/usr/src/kernels/6.0.7-1.el9.elrepo.x86_64'
make[1]: Leaving directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/module'
make -C src
make[1]: Entering directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/src'
Creating Makefile utils.d
Creating Makefile sys.d
Creating Makefile rdsk.d
Creating Makefile rapiddiskd.d
Creating Makefile nvmet.d
Creating Makefile net.d
Creating Makefile main.d
Creating Makefile json.d
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c main.c -o main_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c utils.c -o utils_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c json.c -o json_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c rdsk.c -o rdsk_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c nvmet.c -o nvmet_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c sys.c -o sys_ndebug.o
cc -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o rapiddisk main_ndebug.o utils_ndebug.o json_ndebug.o rdsk_ndebug.o nvmet_ndebug.o sys_ndebug.o  -ljansson -ldevmapper -lpcre2-8
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c rapiddiskd.c -o rapiddiskd_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c json.c -o json-server_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c nvmet.c -o nvmet-server_ndebug.o
cc -Wdate-time -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c net.c -o net_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c rdsk.c -o rdsk-server_ndebug.o
cc -Wdate-time -DSERVER -O2 -Wno-unused-result -flto=auto -ffat-lto-objects -fstack-protector-strong -c sys.c -o sys-server_ndebug.o
cc -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o rapiddiskd rapiddiskd_ndebug.o utils_ndebug.o json-server_ndebug.o nvmet-server_ndebug.o net_ndebug.o rdsk-server_ndebug.o sys-server_ndebug.o  -ljansson -ldevmapper -lpcre2-8 -lmicrohttpd
make[1]: Leaving directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/src'
make -C test
make[1]: Entering directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/test'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/shub/rpmbuild/BUILD/rapiddisk-8.2.0/test'
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%install): /bin/sh -e /var/tmp/rpm-tmp.E2PX9O
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ '[' /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64 '!=' / ']'
+ rm -rf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
++ dirname /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ mkdir -p /home/shub/rpmbuild/BUILDROOT
+ mkdir /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ cd rapiddisk-8.2.0
+ rm -rf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ cd src
+ make DESTDIR=/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/ install
Installing all rapiddisk rapiddiskd binary files.
install -D -m 755 -t /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//sbin rapiddisk rapiddiskd
+ cd ../doc
+ make DESTDIR=/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/ install
Nothing to do.
Installing rapiddisk man pages.
mkdir -pv -m 755 /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man/man1
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man/man1'
cp rapiddisk*.1 /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//usr/share/man/man1/
+ cd ../conf
+ make DESTDIR=/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/ install
Nothing to do.
Installing rapiddisk high availability and rapiddiskd daemon files.
install -D -m 755 -t /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//etc/rapiddisk/ rapiddisk.*
install -D -m 755 -t /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64//etc/systemd/system rapiddiskd.*
+ cd ../module
+ mkdir -pv /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src'
mkdir: created directory '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/'
+ cp -v dkms.conf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
'dkms.conf' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/dkms.conf'
+ cp -v rapiddisk.c rapiddisk-cache.c /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
'rapiddisk.c' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/rapiddisk.c'
'rapiddisk-cache.c' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/rapiddisk-cache.c'
+ cp -v Makefile /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/
'Makefile' -> '/home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64/usr/src/rapiddisk-8.2.0/Makefile'
+ '[' '%{buildarch}' = noarch ']'
+ QA_CHECK_RPATHS=1
+ case "${QA_CHECK_RPATHS:-}" in
+ /usr/lib/rpm/check-rpaths
+ /usr/lib/rpm/check-buildroot
+ /usr/lib/rpm/redhat/brp-ldconfig
+ /usr/lib/rpm/brp-compress
+ /usr/lib/rpm/brp-strip /usr/bin/strip
+ /usr/lib/rpm/brp-strip-comment-note /usr/bin/strip /usr/bin/objdump
+ /usr/lib/rpm/redhat/brp-strip-lto /usr/bin/strip
+ /usr/lib/rpm/brp-strip-static-archive /usr/bin/strip
+ /usr/lib/rpm/redhat/brp-python-bytecompile '' 1 0
+ /usr/lib/rpm/brp-python-hardlink
+ /usr/lib/rpm/redhat/brp-mangle-shebangs
*** WARNING: ./etc/systemd/system/rapiddiskd.service is executable but has no shebang, removing executable bit
mangling shebang in /etc/rapiddisk/rapiddisk.sh.rgmanager from /bin/bash to #!/usr/bin/bash
mangling shebang in /etc/rapiddisk/rapiddisk.sh.pacemaker from /bin/bash to #!/usr/bin/bash
Processing files: rapiddisk-8.2.0-1.x86_64
Provides: rapiddisk = 8.2.0-1 rapiddisk(x86-64) = 8.2.0-1
Requires(interp): /bin/sh /bin/sh /bin/sh
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires(post): /bin/sh
Requires(preun): /bin/sh
Requires(postun): /bin/sh
Requires: /usr/bin/bash libc.so.6()(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3)(64bit) libc.so.6(GLIBC_2.34)(64bit) libc.so.6(GLIBC_2.4)(64bit) libc.so.6(GLIBC_2.7)(64bit) libdevmapper.so.1.02()(64bit) libdevmapper.so.1.02(Base)(64bit) libdevmapper.so.1.02(DM_1_02_97)(64bit) libjansson.so.4()(64bit) libjansson.so.4(libjansson.so.4)(64bit) libmicrohttpd.so.12()(64bit) libpcre2-8.so.0()(64bit) rtld(GNU_HASH)
Processing files: rapiddisk-utils-8.2.0-1.x86_64
Provides: rapiddisk-utils = 8.2.0-1 rapiddisk-utils(x86-64) = 8.2.0-1
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires: /usr/bin/bash libc.so.6()(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3)(64bit) libc.so.6(GLIBC_2.34)(64bit) libc.so.6(GLIBC_2.4)(64bit) libc.so.6(GLIBC_2.7)(64bit) libdevmapper.so.1.02()(64bit) libdevmapper.so.1.02(Base)(64bit) libdevmapper.so.1.02(DM_1_02_97)(64bit) libjansson.so.4()(64bit) libjansson.so.4(libjansson.so.4)(64bit) libmicrohttpd.so.12()(64bit) libpcre2-8.so.0()(64bit) rtld(GNU_HASH)
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
Wrote: /home/shub/rpmbuild/RPMS/x86_64/rapiddisk-utils-8.2.0-1.x86_64.rpm
Wrote: /home/shub/rpmbuild/RPMS/x86_64/rapiddisk-8.2.0-1.x86_64.rpm
Executing(%clean): /bin/sh -e /var/tmp/rpm-tmp.r061qJ
+ umask 022
+ cd /home/shub/rpmbuild/BUILD
+ cd rapiddisk-8.2.0
+ rm -rf /home/shub/rpmbuild/BUILDROOT/rapiddisk-8.2.0-1.x86_64
+ RPM_EC=0
++ jobs -p
+ exit 0
shub@localhost:~/rpmbuild$
```

And it worked (I hope). The resulting rpms have some problems installing, though.

The new deps to add to `pkg/rpm/rapiddisk.spec.rhel` should be:

```pcre2-utf16,pcre2-utf32,device-mapper-libs```

in the `Requires` field, and

```pcre2-devel,device-mapper-devel```

in the `BuildRequires` field.

`libmicrohttpd12` should be changed to `libmicrohttpd` under Centos 9, too.

But this is totally new to me, a check from you is needed...

This is the diff:

```diff
diff --git a/pkg/rpm/rapiddisk.spec.rhel b/pkg/rpm/rapiddisk.spec.rhel
index a9706f5..d4ce1a7a 100644
--- a/pkg/rpm/rapiddisk.spec.rhel
+++ b/pkg/rpm/rapiddisk.spec.rhel
@@ -6,8 +6,8 @@ License: General Public License Version 2
 Group: Applications/System
 Source: %{name}-%{version}.tar.gz
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
-Requires: gcc,make,kernel-headers,kernel-devel,dkms,jansson,libmicrohttpd12
-BuildRequires: kernel-headers,kernel-devel,gcc,make,jansson-devel,libmicrohttpd-devel
+Requires: gcc,make,kernel-headers,kernel-devel,dkms,jansson,libmicrohttpd,pcre2-utf16,pcre2-utf32,device-mapper-libs
+BuildRequires: kernel-headers,kernel-devel,gcc,make,jansson-devel,libmicrohttpd-devel,pcre2-devel,device-mapper-devel

 %description
 The RapidDisk software defined advanced RAM drive and storage
@@ -28,7 +28,7 @@ RapidDisk enabled storage volumes and accessing them either
 locally or across an NVMe Target network.

 %prep
-%setup -T -D -n rapiddisk
+%setup -q

 %build
 make
@@ -130,7 +130,7 @@ rm -rf %{buildroot}
 - utility: Restructured userspace CLI
 - test: Restructured and improved test framework
 - misc: Code / documentation cleanup
-* Thu May 26 2020 Petros Koutoupis <petros@petroskoutoupis.com>
+* Tue May 26 2020 Petros Koutoupis <petros@petroskoutoupis.com>
 - kernel: added support for 5.7 kernel
 - utility: fixed GCC compilation warnings
 - utility: code style cleanup

```

I did not checked the `pkg/rpm/rapiddisk.spec.sles` file.

matteotenca commented:
> > And it worked (I hope). The resulting rpms have some problems installing, though.
> 
> What were the errors or problems?

Here is an example:

```bash
shub@localhost:~/rpmbuild/RPMS/x86_64$ sudo rpm --verbose --install rapiddisk-8.2.0-1.x86_64.rpm
Verifying packages...
Preparing packages...
rapiddisk-8.2.0-1.x86_64
Deprecated feature: REMAKE_INITRD
Creating symlink /var/lib/dkms/rapiddisk/8.2.0/source -> /usr/src/rapiddisk-8.2.0
Sign command: /lib/modules/6.0.7-1.el9.elrepo.x86_64/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub
Deprecated feature: REMAKE_INITRD

Building module:
Cleaning build area...
make -j3 KERNELRELEASE=6.0.7-1.el9.elrepo.x86_64 -C /lib/modules/6.0.7-1.el9.elrepo.x86_64/build M=/var/lib/dkms/rapiddisk/8.2.0/build...
Signing module /var/lib/dkms/rapiddisk/8.2.0/build/rapiddisk.ko
Signing module /var/lib/dkms/rapiddisk/8.2.0/build/rapiddisk-cache.ko
Cleaning build area...
Deprecated feature: REMAKE_INITRD

rapiddisk.ko.xz:
Running module version sanity check.
Error! Module version 8.2.0 for rapiddisk.ko.xz
is not newer than what is already found in kernel 6.0.7-1.el9.elrepo.x86_64 (8.2.0).
You may override by specifying --force.

rapiddisk-cache.ko.xz:
Running module version sanity check.
Error! Module version 8.2.0 for rapiddisk-cache.ko.xz
is not newer than what is already found in kernel 6.0.7-1.el9.elrepo.x86_64 (8.2.0).
You may override by specifying --force.
Adding any weak-modules
depmod.....
/var/tmp/rpm-tmp.NfS3DO: line 4: /etc/sysconfig/modules/rapiddisk.modules: No such file or directory
/var/tmp/rpm-tmp.NfS3DO: line 5: /etc/sysconfig/modules/rapiddisk.modules: No such file or directory
chmod: cannot access '/etc/sysconfig/modules/rapiddisk.modules': No such file or directory
sh: /etc/sysconfig/modules/rapiddisk.modules: No such file or directory
warning: %post(rapiddisk-8.2.0-1.x86_64) scriptlet failed, exit status 127
shub@localhost:~/rpmbuild/RPMS/x86_64$
```

There is an issue when trying to reinstall the rpm, too:

```bash
shub@localhost:~/rpmbuild/RPMS/x86_64$ sudo dnf remove rapiddisk-8.2.0-1.x86_64
Dependencies resolved.
==================================================================================================================================
 Package                         Architecture                 Version                        Repository                      Size
==================================================================================================================================
Removing:
 rapiddisk                       x86_64                       8.2.0-1                        @@System                       232 k

Transaction Summary
==================================================================================================================================
Remove  1 Package

Freed space: 232 k
Is this ok [y/N]: y
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                          1/1
  Running scriptlet: rapiddisk-8.2.0-1.x86_64                                                                                 1/1
Deprecated feature: REMAKE_INITRD
Deprecated feature: REMAKE_INITRD

  Erasing          : rapiddisk-8.2.0-1.x86_64                                                                                 1/1
  Running scriptlet: rapiddisk-8.2.0-1.x86_64                                                                                 1/1
  Verifying        : rapiddisk-8.2.0-1.x86_64                                                                                 1/1

Removed:
  rapiddisk-8.2.0-1.x86_64

Complete!
shub@localhost:~/rpmbuild/RPMS/x86_64$ sudo rpm --verbose --install rapiddisk-8.2.0-1.x86_64.rpm
Verifying packages...
Preparing packages...
rapiddisk-8.2.0-1.x86_64
Error! DKMS tree already contains: rapiddisk-8.2.0
You cannot add the same module/version combo more than once.
Deprecated feature: REMAKE_INITRD
Module rapiddisk/8.2.0 already built for kernel 6.0.7-1.el9.elrepo.x86_64 (x86_64), skip. You may override by specifying --force.
Deprecated feature: REMAKE_INITRD
Module rapiddisk/8.2.0 already installed on kernel 6.0.7-1.el9.elrepo.x86_64 (x86_64), skip. You may override by specifying --force.
/var/tmp/rpm-tmp.6G6Uej: line 4: /etc/sysconfig/modules/rapiddisk.modules: No such file or directory
/var/tmp/rpm-tmp.6G6Uej: line 5: /etc/sysconfig/modules/rapiddisk.modules: No such file or directory
chmod: cannot access '/etc/sysconfig/modules/rapiddisk.modules': No such file or directory
sh: /etc/sysconfig/modules/rapiddisk.modules: No such file or directory
warning: %post(rapiddisk-8.2.0-1.x86_64) scriptlet failed, exit status 127
shub@localhost:~/rpmbuild/RPMS/x86_64$
```

pkoutoupis commented:
OK. That is because the directory `/etc/sysconfig/modules/` is not present. The spec files need a `mkdir -pv` in the %post section:
```
mkdir -pv /etc/sysconfig/modules/
echo -ne "#!/bin/sh\nmodprobe rapiddisk max_sectors=2048 nr_requests=1024 2>&1 >/dev/null" > /etc/sysconfig/modules/rapiddisk.modules
echo -ne "\nmodprobe rapiddisk-cache 2>&1 >/dev/null" >> /etc/sysconfig/modules/rapiddisk.modules
chmod +x /etc/sysconfig/modules/rapiddisk.modules
```

In my environment, this directory is present for one reason or another.

I cannot explain the dkms errors yet. The %preun should have cleaned up the build.

But let me try it in my environment.

pkoutoupis commented on src/main.c:
```diff
@@ -94,326 +98,405 @@ int exec_cmdline_arg(int argcin, char *argvin[])
-			sprintf(device, "%s", optarg);
+			case 'h':
+				printf("%s", header);
+				online_menu(PROCESS);
```
I still prefer the use of argvin[0] for one reason:
If someone changes the binary name to (for example) `rapiddisk-util`, the online_menu will show examples with the executed name.

matteotenca replied:
The problem rises when the specified command argument is something like this, and the messages get messed up:

```bash
shub@localhost:/tmp/rapiddisk-8.2.0/src$ sudo ../../test/.././test/subdir/.././subdir/bin/./testbin/rapiddisk -h
../../test/.././test/subdir/.././subdir/bin/./testbin/rapiddisk 8.2.0
Copyright 2011 - 2022 Petros Koutoupis

../../test/.././test/subdir/.././subdir/bin/./testbin/rapiddisk is an administration tool to manage RapidDisk RAM disk devices and
        RapidDisk-Cache mappings.

Usage: ../../test/.././test/subdir/.././subdir/bin/./testbin/rapiddisk [ -h | -v ] function [ parameters ]

Description:
        ../../test/.././test/subdir/.././subdir/bin/./testbin/rapiddisk is a RapidDisk module management tool to manage RapidDisk
        RAM disk devices. Dynamically create, remove, resize RAM volumes and if
        desired, map or unmap them as a cache volume to any block device.
[...[
```

If you want I rollback the change!

pkoutoupis replied:
No. Keep the change then. You make a great point.

pkoutoupis commented on src/main.c:
```diff
@@ -94,326 +98,405 @@ int exec_cmdline_arg(int argcin, char *argvin[])
-		}
-		online_menu(argvin[0]);
-		break;
+			online_menu(PROCESS);
```
See other note on argvin[0].

pkoutoupis commented on src/main.c:
```diff
@@ -94,326 +98,405 @@ int exec_cmdline_arg(int argcin, char *argvin[])
+			default:
+			case '?':
+				printf("%s", header);
+				online_menu(PROCESS);
```
See other note on argvin[0].

pkoutoupis commented:
@matteotenca Did you have any new commits to push? I believe that we are almost there. I took a few [cosmetic] notes to update some documentation and focus on the nvmet.c area plus the rpm spec file but I really want to merge what we have here. It looks really good. Thoughts?

matteotenca commented:
> @matteotenca Did you have any new commits to push?

I started to hammer the spec file, `pkg/rpm/rapiddisk.spec.rhel`, I ended up with a brand new one, I just committed it. I suppose the `.sles` version must be updated too, but I don't know the differences between the two.


> I believe that we are almost there. I took a few [cosmetic] notes to update some documentation and focus on the nvmet.c area plus the rpm spec file but I really want to merge what we have here. It looks really good. Thoughts?

Yes, we are almost done - I believe a further short-term release will be necessary to fix some bugs that, given the size of the PR, are quite inevitable... Better to release this one and fix bugs as they come up later, with smaller PRs and other small releases.

Regards! (and Merry Christmas!!)

pkoutoupis commented:
@matteotenca Just confirming: Do you feel good about merging now? And again, thank you for the great work and investment into this project.

And Merry Christmas to you as well! :-)

matteotenca commented:
> @matteotenca Just confirming: Do you feel good about merging now?

@pkoutoupis Just some points:

1. The `-q` option in the main `rapiddisk` tool is not documented on the help message. If you provide me with a nice definition, I will commit the modification.
2. If you can, take a look at the issue #142 I just opened, maybe we can fix it before merging.
3. It would be nice if you can double check the new `pkg/rpm/rapiddisk.spec.rhel`, since it's my first attempt to write/correct a `spec` file for `rpmbuild`. Furthermore, the `pkg/rpm/rapiddisk.spec.sles` is still untouched - I don't know which distribution it is for - maybe Slackware?
4. The changelogs must be updated before release

That said, I feel good about merging. The new version of `rapiddisk-on-boot` is near to be ready, but I believe we should add it in the next update. It works, but no documentation is ready and I have to adapt it to run under Centos (`drakut`). Not much of an issue, but I still have to do it.

> And again, thank you for the great work and investment into this project.

You're welcome, I am very happy to provide you with some help, and that you're satisfied with my code!

Regards

pkoutoupis commented:
OK. I will merge tomorrow.
Replies to your notes:
Numbers 1 and 4 are already on my list.
Number 2 seems like it has been there and shouldn't hold up the merge. I will dive into this shortly. Thank you!
Number 3 will be looked into more thoroughly tomorrow because I have it in my list to fix the sles version (SUSE Linux Enterprise Server).

pkoutoupis commented:
I will fix it but a couple of adjustments with the SPEC file:
1. Added a `%global debug_package %{nil}` before the `%prep` section because of some empty debug file errors.
2. Had to fix the date for May 25 2018. Not sure how this slipped past me before.

Anyway, I will be merging this and will start a new branch to cover all of the little things discussed earlier. Thank you again!
